### PR TITLE
Add contiv networking support (https://github.com/kubernetes/contrib/issues/577)

### PR DESCRIPTION
**Pull request by jojimt**
This PR brings in ansible changes needed to setup kubernetes with Contiv networking as outlined in https://github.com/kubernetes/contrib/issues/577.

Summary of changes:
1) Added a new role (contiv) that holds all contiv specific config code
2) Minor tweaks to etcd setup so contiv can access etcd from any node in proxy-mode
3) Exposed a few options via variables, with default values set so the default behavior will be unchanged.
4) Added missing "validate_certs: False" for a couple of addon file fetches.

### CODE REVIEW

File: ansible/cluster.yml [hunk 1]
```diff
@@ -65,6 +71,7 @@
 # install kubernetes on the nodes
 - hosts: nodes
   sudo: yes
+  serial: 5
```
Comment (danehans) [inline ansible/cluster.yml]: Is this required for adding contiv support? If not, can you please create a separate pr for the patch?

File: ansible/group_vars/all.yml [hunk 2]
```diff
@@ -96,3 +96,21 @@ dns_replicas: 1
 
 # See kube documentation for apiserver runtime config options.  Example below enables HPA, deployments features.
 # apiserver_extra_args: "--runtime-config=extensions/v1beta1/deployments=true"
+
```
Comment (danehans) [inline ansible/group_vars/all.yml]: lines 99-101 already exist and should not be in this pr: https://github.com/kubernetes/contrib/blob/master/ansible/group_vars/all.yml#L40-L41

File: ansible/group_vars/all.yml [hunk 3]
```diff
@@ -96,3 +96,21 @@ dns_replicas: 1
...
+# The port that the Kubernetes apiserver component listens on.
+kube_master_api_port: 443
+
+# This directory is where all the additional config stuff goes
+# the kubernetes normally puts in /srv/kubernets.
+# This puts them in a sane location.
+# Editting this value will almost surely break something. Don't
+# change it. Things like the systemd scripts are hard coded to
+# look in here. Don't do it.
+kube_config_dir: /etc/kubernetes
```
Comment (danehans) [inline ansible/group_vars/all.yml]: This already exists: roles/kubernetes/defaults/main.yml
kube_config_dir: /etc/kubernetes

We do not want to expose this in group_vars.

File: ansible/cluster.yml (hunk 1 above)
Comment (jojimt) [inline ansible/cluster.yml]: Good point. This is only required when the cluster has many nodes (I tested with 20). If it will be helpful, I can take it out of this PR.

File: ansible/group_vars/all.yml [hunk 4]
```diff
@@ -96,3 +96,21 @@ dns_replicas: 1
...
+
+# This is where all the cert scripts and certs will be located
+kube_cert_dir: "{{ kube_config_dir }}/certs"
```
Comment (danehans) [inline ansible/group_vars/all.yml]: Same feedback as above. From what I gather, nothing should change in this file except adding contiv to this line: https://github.com/kubernetes/contrib/blob/master/ansible/group_vars/all.yml#L49

File: ansible/roles/contiv/defaults/main.yaml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+# Defaults to the address used by ansible to reach the node.
+# If using vagrant, this will need to be overriden or edited here.
+contiv_control_ip: "{{ hostvars[inventory_hostname]['ansible_default_ipv4']['address'] }}"
+
+# This is the host interface that contiv uses for all inter-host tenant traffic. Typically set
+# in the inventory file.
+#contiv_network_if: ethX
+
+# This is the path where contiv binaries are kept
+contiv_bin_path: ~/go/bin
```
Comment (danehans) [inline ansible/roles/contiv/defaults/main.yaml]: Can centavo follow the same bin_dir pattern found here: roles//common/defaults/main.yaml:
bin_dir: /usr/bin

bin_dir fact is used for supporting multiple distros

File: ansible/roles/contiv/defaults/main.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+# if you are using an external router for all routing, you should select bridge here
+contiv_fwd_mode: routing
+
+# contiv_def_subnet is the subnet pods will be assigned by default (i.e. pod spec does not
+# specify a network)
+contiv_def_subnet: "20.1.1.0/24"
+contiv_defnet_gw: 20.1.1.254
+
+# The following are aci specific parameters. These must be set only if aci integration
+# is desired. Otherwise, you can ignore these.
```
Comment (danehans) [inline ansible/roles/contiv/defaults/main.yaml]: Do you have a link that explains aci integration in more detail? If so, pls add it as a remark.

File: ansible/roles/contiv/files/contiv_cni.conf [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "cniVersion": "0.1.0",
```
Comment (danehans) [inline ansible/roles/contiv/files/contiv_cni.conf]: can u parameterize the chi version instead of hard coding. Throw the param/value in defaults/main

File: ansible/group_vars/all.yml (hunk 3 above)
Comment (jojimt) [inline ansible/group_vars/all.yml]: Actually, this is needed for other roles. That's why it is moved to group_vars now.

File: ansible/roles/contiv/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+# initial set up of netplugin
+- include: netplugin.yml
+  when: contiv_role == "netplugin"
+
+# clean restart of netplugin
+- include: restart_netplugin.yml tags=contiv_restart
+  when: contiv_role == "netplugin"
+
+# copy/load images used for demo
```
Comment (danehans) [inline ansible/roles/contiv/tasks/main.yml]: I would like to see the demo stuff removed. If the demo stuff ids used for testing, then I'm all for that. Lets just change the name accordingly and make sure their is a test clean-up after a successful test.

File: ansible/roles/contiv/tasks/netmaster.yml [hunk 9]
```diff
@@ -0,0 +1,29 @@
+
+- name: copy netmaster exec file
+  copy: src={{ contiv_bin_path }}/netmaster dest=/usr/bin/netmaster mode=0755
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netmaster.yml]: We can't have /usr/bin/ hardcoded b/c /usr/bin/ is RO on coreos. Can you follow the existing design pattern for copying bins?

File: ansible/group_vars/all.yml (hunk 4 above)
Comment (jojimt) [inline ansible/group_vars/all.yml]: I am installing the certificates on the nodes so the contiv client can access the api server. That's why this is moved to group_vars.

File: ansible/roles/contiv/tasks/netmaster.yml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+- name: copy systemd units for netmaster
+  copy: src=netmaster.service dest=/etc/systemd/system/netmaster.service
+
+- name: setup netmaster host alias
+  lineinfile: dest=/etc/hosts line="{{ contiv_service_vip }} netmaster"
+
+- name: restart netmaster
+  service: name=netmaster state=restarted
+
+- name: wait for netmaster
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netmaster.yml]: Is this needed? I believe service: name=netmaster state=restarted will fail if the restart is not successful. Unusable runs are long, so we want to everything we can to make things quick.

File: ansible/group_vars/all.yml (hunk 2 above)
Comment (jojimt) [inline ansible/group_vars/all.yml]: Not sure why it appears after the rebase. No conflicts are marked either.

File: ansible/roles/contiv/defaults/main.yaml (hunk 5 above)
Comment (jojimt) [inline ansible/roles/contiv/defaults/main.yaml]: Yes, will do.

File: ansible/roles/contiv/tasks/main.yml (hunk 8 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/main.yml]: Yes, it is used for testing. I will update the names.

File: ansible/roles/contiv/tasks/netmaster.yml (hunk 9 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netmaster.yml]: Sure.

File: ansible/roles/contiv/tasks/netmaster.yml (hunk 10 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netmaster.yml]: This is needed for now. I will optimize it to the exact required delay later.

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 11]
```diff
@@ -0,0 +1,81 @@
+- name: ensure pluginDir exists
+  file: path=/usr/libexec/kubernetes/kubelet-plugins/net/exec recurse=yes state=directory
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: We will need to work around this for cores support b/c /usr/libexec is RO.

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+- name: copy contiv cni conf file
+  copy: src=contiv_cni.conf dest=/usr/libexec/kubernetes/kubelet-plugins/net/exec/contiv_cni.conf
+
+- name: ensure cni bin dir exists
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: can the chi bin live with the rest of the bins using the bin_dir variable? I We want to have a single location where we manage bins.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 11 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: Sure, I can do that. However, if I remember right, this path is hard-coded in kubelet.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 12 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: As of now, no.

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+- name: get openstack kilo rpm
+  get_url:
+    url: https://repos.fedorapeople.org/repos/openstack/openstack-kilo/rdo-release-kilo-1.noarch.rpm
+    dest: /tmp/rdo-release-kilo-1.noarch.rpm
+    validate_certs: False
+  environment:
+    http_proxy: "{{ http_proxy|default('') }}"
+    https_proxy: "{{ https_proxy|default('') }}"
+  when: ansible_os_family == "RedHat"
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: Will ^ work using ansible_distribution instead? The rest of the project uses ansible_distribution, so it would be nice to continue using that fact.

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  copy: src={{ contiv_bin_path }}/netplugin dest=/usr/bin/netplugin mode=0755
+
+- name: get openstack kilo rpm
+  get_url:
+    url: https://repos.fedorapeople.org/repos/openstack/openstack-kilo/rdo-release-kilo-1.noarch.rpm
+    dest: /tmp/rdo-release-kilo-1.noarch.rpm
+    validate_certs: False
+  environment:
+    http_proxy: "{{ http_proxy|default('') }}"
+    https_proxy: "{{ https_proxy|default('') }}"
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: I think this also needs no_proxy. See the following for an example:
roles//etcd/tasks/github-release.yml

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+- name: ensure contiv config directory exists
+  file: path=/opt/contiv/config recurse=yes state=directory
+
+- name: setup config for the contiv cni plugin
+  template: src=contiv.cfg.j2 dest=/opt/contiv/config/contiv.json
+
+- name: copy netplugin exec file
+  copy: src={{ contiv_bin_path }}/netplugin dest=/usr/bin/netplugin mode=0755
+
+- name: get openstack kilo rpm
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: This will be another problem for supporting coreOS. Do you mind opening GH Issues for these problem ares so we don't loose track of them?

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 13 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: Sorry, did not understand this comment. Can you explain what you mean by "Will ^ work using ansible_distribution instead?"

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  environment:
+    http_proxy: "{{ http_proxy|default('') }}"
+    https_proxy: "{{ https_proxy|default('') }}"
+  when: ansible_os_family == "RedHat"
+  tags:
+       - ovs_install
+
+- name: install openstack kilo
+  yum: name=/tmp/rdo-release-kilo-1.noarch.rpm state=present
+  when: ansible_os_family == "RedHat"
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: s/ansible_os_family == "RedHat"/ansible_distribution == "CentOS"/g

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  tags:
+       - ovs_install
+
+- name: install ovs
+  environment:
+    http_proxy: "{{ http_proxy|default('') }}"
+    https_proxy: "{{ https_proxy|default('') }}"
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: add no_proxy

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  yum: name=openvswitch state=present
+  when: ansible_os_family == "RedHat"
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: s/ansible_os_family == "RedHat"/ansible_distribution == "CentOS"/g

File: ansible/roles/contiv/tasks/netplugin.yml [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  tags:
+       - ovs_install
+
+- name: install ntp
```
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: Is ntp required for contiv? If not, pls remove. I think it's a good idea to add nt. support to the common role, but that should be done in a separate PR if centavo does not require ntp.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 14 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: Normally, it won't require it. However, I guess, it makes sense to have no_proxy also exported when we export the proxy. I will add it.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 15 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: I don't really know the details of all CoreOS issues. If you're aware of the issue, why don't you open it?

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 19 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: Yes, it is required.

File: ansible/roles/contiv/tasks/restart_netmaster.yml [hunk 20]
```diff
@@ -0,0 +1,35 @@
+- name: stop netmaster
```
Comment (danehans) [inline ansible/roles/contiv/tasks/restart_netmaster.yml]: restart_netmaster.yml performs the same tasks as net master.yml? Is this really needed? Would a handler and notify help? Take a look at the following for add'l info on the handler/notify idea:

roles/docker/handlers/main.yml

File: ansible/roles/contiv/tasks/restart_netplugin.yml [hunk 21]
```diff
@@ -0,0 +1,71 @@
+# stop services
```
Comment (danehans) [inline ansible/roles/contiv/tasks/restart_netplugin.yml]: same with restart_netmaster.yml. It would be really nice if we can get away from having to duplicate all the code in the start netplugin/master just to restart the services.

File: ansible/roles/contiv/tasks/restart_netmaster.yml (hunk 20 above)
Comment (jojimt) [inline ansible/roles/contiv/tasks/restart_netmaster.yml]: Yes, this is needed because sometimes we want to explicitly restart the networking.

File: ansible/roles/etcd/templates/etcd.conf.j2 [hunk 22]
```diff
@@ -1,40 +1,48 @@
 {% macro initial_cluster() -%}
 {% for host in groups[etcd_peers_group] -%}
 {% if loop.last -%}
-{{ hostvars[host]['ansible_hostname'] }}={{ etcd_peer_url_scheme }}://{{ hostvars[host].ansible_default_ipv4.address }}:{{ etcd_peer_port }}
+{{ hostvars[host]['ansible_hostname'] }}={{ etcd_peer_url_scheme }}://{{ hostvars[host]['ansible_fqdn'] }}:{{ etcd_peer_port }}
```
Comment (danehans) [inline ansible/roles/etcd/templates/etcd.conf.j2]: you need to use ansible_default_ipv4.address and not ansible_fqdn

File: ansible/roles/etcd/templates/etcd.conf.j2 [hunk 23]
```diff
@@ -1,40 +1,48 @@
...
 {%- else -%}
-{{ hostvars[host]['ansible_hostname'] }}={{ etcd_peer_url_scheme }}://{{ hostvars[host].ansible_default_ipv4.address }}:{{ etcd_peer_port }},
+{{ hostvars[host]['ansible_hostname'] }}={{ etcd_peer_url_scheme }}://{{ hostvars[host]['ansible_fqdn'] }}:{{ etcd_peer_port }},
```
Comment (danehans) [inline ansible/roles/etcd/templates/etcd.conf.j2]: you need to use ansible_default_ipv4.address and not ansible_fqdn

File: ansible/roles/etcd/templates/etcd.conf.j2 [hunk 24]
```diff
@@ -1,40 +1,48 @@
...
 {%- endif -%}
 {% endfor -%}
 {% endmacro -%}
 
-{% if groups[etcd_peers_group] and groups[etcd_peers_group] | length > 1 %}
+{% if groups[etcd_peers_group] and groups[etcd_peers_group] | length > 0 %}
```
Comment (danehans) [inline ansible/roles/etcd/templates/etcd.conf.j2]: Is length > 0 required for contiv support. If so, please explain.

File: ansible/roles/etcd/templates/etcd.conf.j2 [hunk 25]
```diff
@@ -1,40 +1,48 @@
...
 #ETCD_SNAPSHOT_COUNTER="10000"
 #ETCD_HEARTBEAT_INTERVAL="100"
 #ETCD_ELECTION_TIMEOUT="1000"
-ETCD_LISTEN_CLIENT_URLS="{{ etcd_listen_client_urls }}"
 #ETCD_MAX_SNAPSHOTS="5"
 #ETCD_MAX_WALS="5"
 #ETCD_CORS=""
 
-{% if groups[etcd_peers_group] and groups[etcd_peers_group] | length > 1 %}
+
```
Comment (danehans) [inline ansible/roles/etcd/templates/etcd.conf.j2]: Are you adding support for etcd ha below? If so, @adamschaub is working on that in a separate pr. Please remove. I'm a bit confused on why all these etc changes are required for Contiv.

File: ansible/roles/etcd/templates/etcd.conf.j2 (hunk 23 above)
Comment (jojimt) [inline ansible/roles/etcd/templates/etcd.conf.j2]: Why is that? In my set up, ansible_default_ipv4.address actually gives a syntax error. It can be ['ansible_default_ipv4'].address. But ansible_fqdn works just fine. Can you explain what is the issue here?

File: ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml [hunk 26]
```diff
@@ -13,6 +13,7 @@
     url=https://raw.githubusercontent.com/GoogleCloudPlatform/kubernetes/master/cluster/addons/cluster-monitoring/influxdb/{{ item }}
     dest="{{ local_temp_addon_dir }}/cluster-monitoring/{{ item }}.j2"
     force=yes
+    validate_certs=False
```
Comment (danehans) [inline ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml]: pls remove and do an overall scrub on changes that are outside the scope of what's needed to add Contiv support.

File: ansible/roles/etcd/templates/etcd.conf.j2 (hunk 24 above)
Comment (jojimt) [inline ansible/roles/etcd/templates/etcd.conf.j2]: Yes, because we run etcd in proxy mode on all the nodes and non-proxy mode on a single master.

File: ansible/roles/kubernetes-addons/tasks/dns.yml [hunk 27]
```diff
@@ -13,6 +13,7 @@
     url=https://raw.githubusercontent.com/GoogleCloudPlatform/kubernetes/master/cluster/addons/dns/skydns-rc.yaml.in
     dest="{{ local_temp_addon_dir }}/dns/skydns-rc.yaml.j2"
     force=yes
+    validate_certs=False
```
Comment (danehans) [inline ansible/roles/kubernetes-addons/tasks/dns.yml]: pls remove and do an overall scrub on changes that are outside the scope of what's needed to add Contiv support. If their is an issue with cert validation, pls open a GH issue and provide an explanation or a a PR that explains and fixes the issue.

File: ansible/roles/kubernetes-addons/tasks/dns.yml [hunk 28]
```diff
@@ -38,6 +39,7 @@
     url=https://raw.githubusercontent.com/GoogleCloudPlatform/kubernetes/master/cluster/addons/dns/skydns-svc.yaml.in
     dest="{{ local_temp_addon_dir }}/dns/skydns-svc.yaml.j2"
     force=yes
+    validate_certs=False
```
Comment (danehans) [inline ansible/roles/kubernetes-addons/tasks/dns.yml]: pls remove and do an overall scrub on changes that are outside the scope of what's needed to add Contiv support. If their is an issue with cert validation, pls open a GH issue and provide an explanation or a a PR that explains and fixes the issue.

File: ansible/roles/etcd/templates/etcd.conf.j2 (hunk 25 above)
Comment (jojimt) [inline ansible/roles/etcd/templates/etcd.conf.j2]: No, I am only adding proxy mode support as contiv requires it.

File: ansible/roles/kubernetes/tasks/secrets.yml [hunk 29]
```diff
@@ -39,10 +39,42 @@
   set_fact:
```
Comment (danehans) [inline ansible/roles/kubernetes/tasks/secrets.yml]: All these changes seems outside the scope of Contiv support. If I'm wrong, please explain.

File: ansible/roles/node/defaults/main.yml [hunk 30]
```diff
@@ -1,2 +1,4 @@
 kubelet_working_dir: /var/lib/kubelet
 localBuildOutput: ../../_output/local/go/bin
+# Specifies the number of pods allowed on a node.
+max_pods: "40"
```
Comment (danehans) [inline ansible/roles/node/defaults/main.yml]: This should be separate of Contiv support, right?

File: ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml (hunk 26 above)
Comment (jojimt) [inline ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml]: The problem is that the run fails without this. When I see that other similar code has this set, it seems reasonable.

File: ansible/roles/node/templates/kubelet.j2 [hunk 31]
```diff
@@ -14,12 +14,16 @@ KUBELET_HOSTNAME="--hostname-override={{ inventory_hostname }}"
 KUBELET_API_SERVER="--api-servers=https://{{ groups['masters'][0] }}:{{ kube_master_api_port }}"
 
 {% set kubelet_options = [] -%}
+{{ kubelet_options.append('--max-pods=' + max_pods|string)|default('', true) -}}
```
Comment (danehans) [inline ansible/roles/node/templates/kubelet.j2]: This should be separate of Contiv support, right?

File: ansible/roles/kubernetes-addons/tasks/dns.yml (hunk 27 above)
Comment (jojimt) [inline ansible/roles/kubernetes-addons/tasks/dns.yml]: Same response as above.

File: ansible/roles/kubernetes/tasks/secrets.yml (hunk 29 above)
Comment (jojimt) [inline ansible/roles/kubernetes/tasks/secrets.yml]: No, this is required so we can access the api server using certificate auth.

File: ansible/roles/node/defaults/main.yml (hunk 30 above)
Comment (jojimt) [inline ansible/roles/node/defaults/main.yml]: Sure, I can remove this.

File: ansible/cluster.yml (hunk 1 above)
Comment (danehans) [inline ansible/cluster.yml]: We need the PR to only contain changes that are required for adding Contiv. Any additional changes that contrib will benefit from, pls create separate pr's.

File: ansible/group_vars/all.yml (hunk 2 above)
Comment (danehans) [inline ansible/group_vars/all.yml]: Weird, take them out then. We need to make sure this PR does not change these params.

File: ansible/group_vars/all.yml (hunk 3 above)
Comment (danehans) [inline ansible/group_vars/all.yml]: Please explain, if the centavo roles needs this, then the centavo role should have a dependency on the cabernets role. Checkout how open contrail does this:
```
$ cat roles/opencontrail/meta/main.yml 
---
dependencies:
  - { role: common }
  - { role: kubernetes }
```

File: ansible/group_vars/all.yml (hunk 4 above)
Comment (danehans) [inline ansible/group_vars/all.yml]: If the Contiv roles needs this, then the Contiv role should have a dependency on the Kubernetes role. Checkout how opencontrail does this:
```
$ cat roles/opencontrail/meta/main.yml 
---
dependencies:
  - { role: common }
  - { role: kubernetes }
```

We def do not want to expose these dir's in group_vars.

File: ansible/roles/contiv/tasks/main.yml (hunk 8 above)
Comment (danehans) [inline ansible/roles/contiv/tasks/main.yml]: OK, I think it's also important to include a task that cleans-up after the test.

File: ansible/group_vars/all.yml (hunk 3 above)
Comment (jojimt) [inline ansible/group_vars/all.yml]: I need to template the location of certificates so contiv client can access the kube api server. Please see ansible/roles/contiv/templates/contiv.cfg.j2. That's why these variables need to be exposed. What would be the other way to do that?

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 11 above)
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: Right, but we will need to use an Ansible role outside of contrib that creates the dir using cloud-init to support coreOS. In general, we want to eliminate as much hard-coding in tasks as possible.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 15 above)
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: OK, I will.

File: ansible/roles/contiv/tasks/netplugin.yml (hunk 13 above)
Comment (danehans) [inline ansible/roles/contiv/tasks/netplugin.yml]: The rest of the contrib project uses the ansible_distribution fact for OS detection. For example: ansible_distribution == "CentOS"

Can you use ansible_distribution == "CentOS" instead of ansible_os_family == "RedHat" so the Contiv role follows the same design pattern as the rest of the project?

Comment (jojimt) [inline ansible/roles/contiv/tasks/netplugin.yml]: OK, will change that.

File: ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml (hunk 26 above)
Comment (danehans) [inline ansible/roles/kubernetes-addons/tasks/cluster-monitoring.yml]: It looks like you found a bug, where some add ons are missing this. I agree it's needed. For the purposes of keeping the PR clean, add notes to the PR description for changes above and beyond Contiv support. This helps if we need to revert, etc.. for PR's to match what changes they make.

**Comment by danehans** [comment]
@jojimt my vagrant install fails when using Contiv:

```
TASK [contiv : copy netmaster exec file] ***************************************
fatal: [kube-master]: FAILED! => {"changed": false, "failed": true, "msg": "could not find src=/Users/daneyonhansen/go/bin/netmaster"}
```

Seems like you need to create contiv_bin_path before copy netmaster exec file. Are you using contrib vagrant to test or are you using a different test environment?

**Comment by jojimt** [comment]
Yes, you need to create that directory on a local machine. I use a shell script to do that before starting vagrant.

**Comment by danehans** [comment]
I think the ansible contiv role should be doing that. The other roles provide similar functionality.

**Comment by danehans** [comment]
@jojimt can you send me the shell script so I can replicate and possibly add it to vagrant. As I mentioned above though, I believe the ansible role should be creating that bin dir.

**Comment by jojimt** [comment]
Here is an example: https://github.com/contiv/netplugin/blob/master/vagrant/k8s/setup_cluster.sh

From what I understand, even in k8s case, the ansible does not create the localBinOutput directory. It needs to be created with appropriate binaries before invoking ansible. I just followed the same pattern.

**Comment by danehans** [comment]
@eparis you can close this since https://github.com/kubernetes/contrib/pull/783 merged.
